### ISSUE TITLE
Googlebot can't crawl commonwealth.im

### ISSUE THREAD

**Issue by timolegros** [hicommonwealth/commonwealth#6165]
## Describe the bug

Our prerender service is not set up correctly so any bot with the Googlebot user agent (web crawler) can't crawl our website. This is potentially heavily impacting our SEO.

Specifically, we use:
```
import prerenderNode from 'prerender-node';
setupPrerenderServer();
app.use(prerenderNode.set('prerenderServiceUrl', 'http://localhost:3000'));
```

## To Reproduce

1. Follow instructions to change your user agent to Googlebot: https://www.searchenginejournal.com/change-user-agent/368448/
2. View commonwealth.im

## Expected behavior

Visiting commonwealth.im with a Googlebot user agent should result in pure HTML being returned (no JS) with no application errors.

## Additional context
- Works ("great") locally, but does not work on Heroku

**Comment by timolegros** [comment]
After testing/tinkering, I was unable to get prerendering to work on Heroku when using the localhost prerender service. I was, however, able to get prerender working using prerender.io. Unless we somehow figure out how to prerender locally on Heroku, I think our options as follows:
- use the paid prerender.io service
- run the service on another Heroku app (or Vultr :/).
- don't prerender

**Comment by kurtisassad** [comment]
####TL;DR
Assuming the pricing is fine, I think that it is good to go with this for now since it doesn't require a lot of work. But I don't have any idea how prerender.io works, so its a little difficult to compare to the fully blown SSR (example questions, What is the latency to the prerender.io server, how long does it take to process the SSRed page, does it SSR dynamic content, is there a caching policy and is it customizable , etc).

#### Alternative
If we went with the full blown SSR route, it would take a ton of work. We would have to:

1. We identify the critical webpages that we would like to SSR
2. We move all necessary global state (zustand, global controllers) for users onto the backend (in our db), so that it is separate from our front end
3. We re-write critical pages that we need SSOed in some meta-framework that allows SSR (next, vite, remix, ...)
4. Spin up a reverse proxy to route these pages to the SSR server

For example it would look something like this:
<img width="565" alt="image" src="https://github.com/hicommonwealth/commonwealth/assets/14794654/f29a8299-3bde-4d77-9aec-af6b2faf28f2">

This comes with some benefits:
* We can gradually perform this migration
* We can use a real SSR framework, one that we have full control over (we can customize what we prerender as well as caching policies)
* This will improve TTFB by 1 round trip time for API calls (~300ms for overseas users)

As well as some downsides:
* We will need to run at minimum 3 servers
* It will take a bunch of time to set up this new flow on heroku
* We will need to refactor out the global state used in these critical pages (any zustand state, any controller usage from the state.ts file)

**Comment by timolegros** [comment]
> #### TL;DR
> Assuming the pricing is fine, I think that it is good to go with this for now since it doesn't require a lot of work. But I don't have any idea how prerender.io works, so its a little difficult to compare to the fully blown SSR (example questions, What is the latency to the prerender.io server, how long does it take to process the SSRed page, does it SSR dynamic content, is there a caching policy and is it customizable , etc).
> 
> #### Alternative (expanding on bullet point 2)
> If we went with the full blown SSR route, it would take a ton of work. We would have to:
> 
> 1. We identify the critical webpages that we would like to SSR
> 2. We move all necessary global state (zustand, global controllers) onto the backend (in our db), so that it is no longer stored on our front end
> 3. We re-write critical pages that we need SSRed in some meta-framework that allows SSR (next, vite, remix, ...)
> 4. Spin up a reverse proxy to route these pages to the SSR server
> 
> For example it would look something like this: <img alt="image" width="843" src="https://private-user-images.githubusercontent.com/14794654/293709336-bea3590b-b6ff-4d0e-854f-c360b94dfd72.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQyMTA5NjAsIm5iZiI6MTcwNDIxMDY2MCwicGF0aCI6Ii8xNDc5NDY1NC8yOTM3MDkzMzYtYmVhMzU5MGItYjZmZi00ZDBlLTg1NGYtYzM2MGI5NGRmZDcyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTAyVDE1NTEwMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRjMmE5MzYzYjYyODEwYTg5NmMxZTMxZGVjYTU1M2U0YjE1MTNlMTFjZWFhN2Y5MmE5ODU4YjdmNWI3NTE2YjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.DGdDHlFYl190zo4R7FAAb5j-lpCDfwxORyx2RxFikCc">
> 
> This comes with some benefits:
> 
> * We can gradually perform this migration (one webpage at a time)
> * We can use a real SSR framework, one that we have full control over (we can customize what we prerender as well as caching policies)
> * This will improve TTFB by 1 round trip time for API calls (~300ms for overseas users)
> 
> As well as some downsides:
> 
> * We will need to run at minimum 3 servers
> * It will take a bunch of time to set up this new flow on heroku
> * We will need to refactor out the global state used in these critical pages (any zustand state, any controller usage from the state.ts file)

Point 2 is different from your alternate solution - technically unrelated and much simpler. Preprender.io provides their prerender server open-source which is so far what we have been running on localhost. The server takes a different approach, rather than moving all state to the backend and all that, it just uses headless chrome to render pages and then removes `script` tags (removing JS) before caching the resulting HTML. So really it's using CSR to create SSR. This is not useful at all for actual users but works well for bots who don't interact with the UI. Your solution is of course all-encompassing since it would work for bots and real users but that's out of scope here.

**Comment by timolegros** [comment]
A couple interesting findings. It seems that prerendering is failing for 2 different reasons in production:
1. there is some kind of issue related to domains
```
Failed fetching custom domain TypeError: Cannot read properties of undefined (reading 'evmTestEnv')
    at o (app.c525cb21.js:125:103593)
    at o.next (<anonymous>)
    at a (app.c525cb21.js:125:104996)
```
2. the local prerender server is failing to parse the html probably because it is failing to render the page. I've dug extensively into the source code of both the prerender middleware and prerender server and I can't see why there would be a difference on Heroku.

That said, even if we forked and fixed the prerender server to work for us on Heroku it is still worth moving to use prerender.io because it means (on top of all the features of prerender.io) that we can remove the google-chrome buildpack and reduce our final slug size to ~280 MB.

**Comment by timolegros** [comment]
Bumping to reflect the amount of time I have spent debugging prerender, researching prerender.io, and learning about our SEO situation (e.g. sitemaps)
